config: support configuration constraints

NyaliaLui opened the pull request:
https://github.com/redpanda-data/redpanda/pull/13942 is stale, use this one instead.

Enable users to set configuration constraints for cluster-level properties. This patch adds support for constraints for a sub-set of cluster-level configs. Specifically, those that are settable via the Kafka CreateTopics API or reported by the DescribeConfigs API. 

Fixes: https://github.com/redpanda-data/redpanda/issues/13571
Fixes: https://github.com/redpanda-data/redpanda/issues/14052

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v23.2.x
- [ ] v23.1.x
- [ ] v22.3.x

## Release Notes

* none

vbotbuildovich commented:
new failures detected in https://buildkite.com/redpanda/redpanda/builds/39495#018b4f98-17c8-431b-9017-c102f18f1788: "rptest.tests.cluster_config_test.ConfigConstraintsTest.test_constraint_configs_persist"

vbotbuildovich commented:
new failures detected in https://buildkite.com/redpanda/redpanda/builds/39495#018b4f9f-ad33-410e-a4f3-9280432f6cbb: "rptest.tests.cluster_config_test.ConfigConstraintsTest.test_constraint_configs_persist"

vbotbuildovich commented:
ducktape was retried in job https://buildkite.com/redpanda/redpanda/builds/39495#018b4f9f-ad36-49e7-8460-23d986d591f4

vbotbuildovich commented:
new failures detected in https://buildkite.com/redpanda/redpanda/builds/39510#018b5068-953e-45ec-9f9b-c68a7e7044d0: "rptest.tests.node_pool_migration_test.NodePoolMigrationTest.test_migrating_redpanda_nodes_to_new_pool.balancing_mode=node_add"

rockwotj commented on src/v/redpanda/application.cc:
```diff
@@ -748,6 +749,13 @@ void application::hydrate_config(const po::variables_map& cfg) {
+    // convert those items to concrete constraint objects and keep them in a
+    // store.
+    ss::smp::invoke_on_all([] {
+        config::shard_local_constraints().load_constraints();
```
Do we need to do this when configuration changes via the Admin API?

NyaliaLui replied:
Yes.

rockwotj reviewed (commented):
Is the actual clamping of topic properties going to come in another PR?

How are we going to handle a constraint that is added and contradicts an existing value? For example, let's say we set `default_topic_replications` to [3, 9] - what happens if the current cluster config value here is set to `1`? Additionally, this is going to clamp topic properties set in the kafka API, what happens if there is an existing topic with RF=1?

BenPope commented on src/v/config/constraint_store.h:
```diff
@@ -0,0 +1,74 @@
+        _constraints.reserve(constraint_args.size());
+        for (const auto& [name, args] : constraint_args) {
+            auto& property = conf.get(name);
+            constraint cons{
+              std::move(args), property.make_constraint_methods()};
+            _constraints[name] = std::move(cons);
```
The move doesn't do anything, and `insert_or_assign` avoids the need for a default constructor.
```suggestion
            constraint cons{args, property.make_constraint_methods()};
            _constraints.insert_or_assign(name, std::move(cons));
```

BenPope commented on src/v/config/constraint_store.h:
```diff
@@ -0,0 +1,74 @@
+        }
+    }
+
+    template<typename Func>
+    void for_each(Func&& f) const {
+        for (const auto& [_, cons] : _constraints) {
+            f(cons);
+        }
+    }
```
Is this used?

NyaliaLui replied:
It's used the patch for the Admin API. I will remove it from this PR.

BenPope commented on src/v/config/property.h:
```diff
@@ -911,4 +956,43 @@ class retention_duration_property final
+        std::move(ld),
+        std::move(constraint_methods_h)) {}
+
+    template<typename Func>
+    helpers::range_values<T> make_range(
+      const std::optional<ss::sstring>& min_str,
+      const std::optional<ss::sstring>& max_str,
+      Func&& type_converter) {
+        auto min = min_str ? std::make_optional(
+                     static_cast<T>(type_converter(*min_str)))
+                           : std::nullopt;
+        auto max = max_str ? std::make_optional(
+                     static_cast<T>(type_converter(*max_str)))
+                           : std::nullopt;
+        return helpers::range_values<T>{std::move(min), std::move(max)};
+    }
```
This looks almost the same as the one in property, which isn't virtual and doesn't rely on any state. It makes me wonder if `make_range` shouldn't just be a free function that can be used from the 3 places it's implemented.

NyaliaLui replied:
`range_values` needs the integral type and the type for each config is captured in the property class that is why `make_range` is a member of property classes.

NyaliaLui replied:
This method will be irrelevant in the next push following a change to integral properties.

BenPope commented on src/v/redpanda/application.cc:
```diff
@@ -748,6 +749,13 @@ void application::hydrate_config(const po::variables_map& cfg) {
+    // store.
+    ss::smp::invoke_on_all([] {
+        config::shard_local_constraints().load_constraints();
+    }).get0();
```
`get0` was deprecated a long time ago.
```suggestion
    }).get();
```

BenPope commented on src/v/tristate.h:
```diff
@@ -122,3 +122,9 @@ class tristate {
     underlying_t _value;
 };
+
+// Factory method so we don't have to code in the type
+template<typename T>
+tristate<T> make_tristate(T val) {
+    return tristate<T>(std::make_optional<T>(val));
+}
```
We have CTAD, so a constructor would also work:
```c++
    constexpr explicit tristate(T t) noexcept
      : _value(std::move(t)) {}
```

BenPope replied:
Ahh, it might create an ambiguity, don't bother spending time fixing that if its the case.

BenPope reviewed (commented):
There's a lot of duplication in the validate and clamp methods, although there are some subtleties across the properties.

NyaliaLui commented:
> > Is the actual clamping of topic properties going to come in another PR?
> 
> Yes #13630

Correct. Clamping comes in a later PR.

> 
> > How are we going to handle a constraint that is added and contradicts an existing value? For example, let's say we set `default_topic_replications` to [3, 9] - what happens if the current cluster config value here is set to `1`? Additionally, this is going to clamp topic properties set in the kafka API, what happens if there is an existing topic with RF=1?
> 
> I believe this is covered in the [RFC](https://docs.google.com/document/d/1M7ZhOA23hiW3RW6HjCEWhe64sWPXGssmkjyxDDDzNYM/edit). Attempting to add a constraint while a topic exists that would violate the constraint would cause the attempt to add the constraint to fail.

Thanks mike. Yes, this is covered in the RFC. Tyler, in your example with RF=1, setting the constraint would fail because the cluster already has a topic that breaks the constraint.

NyaliaLui commented on src/v/config/constraint_store.h:
```diff
@@ -0,0 +1,74 @@
+
+    bool contains(ss::sstring& name) { return _constraints.contains(name); }
+
+    constraint& get(const ss::sstring& name) {
```
> Also where is this method used? It looks like it's not yet?

It is used in unit tests in the following commit. I can move it over.

NyaliaLui commented on src/v/config/constraint_types.h:
```diff
@@ -0,0 +1,155 @@
+    using key_type = ss::sstring;
+
+    ss::sstring name;
+    constraint_type type;
```
> I feel like we could infer this off of the selected options below and then maybe an enable/disable flag to cover the `none` situation.

I am not sure I understand your point here, could you clarify? Remember, users can set restrikt or clamp type for both min/max and enabled constraints.

NyaliaLui commented on src/v/config/validators.cc:
```diff
@@ -164,4 +165,30 @@ validate_audit_event_types(const std::vector<ss::sstring>& vs) {
+              fmt::format("Constraints failure[min > max]: name {}", name));
+        }
+
+        if (
+          (min || max)
+          && (args.enabled && *args.enabled == constraint_enabled_t::yes)) {
+            return std::make_optional<ss::sstring>(fmt::format(
+              "Constraints failure[use 'enabled' or 'min/max' but not both]: "
+              "name {}",
+              name));
+        }
```
I am confused by your question here, could you clarify? Remember, the type does not depend on the options. A user could have restikt and min/max or clamp and min/max. Same thing with enabled.

NyaliaLui commented on src/v/config/constraint_types.cc:
```diff
@@ -0,0 +1,176 @@
+}
+
+constraint_methods::constraint_methods()
+  : validator{std::move(constraint_validator_always_true)}
+  , clamper{std::move(constraint_clamper_noop)}
```
The constructor is so more code is in the impl file instead of the header but I'm also not consistent with that so I will go back to member initialization.

NyaliaLui commented on src/v/config/validators.cc:
```diff
@@ -164,4 +165,30 @@ validate_audit_event_types(const std::vector<ss::sstring>& vs) {
+    constraint_args) {
+    for (const auto& [name, args] : constraint_args) {
+        // Use any integral type since we are checking for min > max.
+        auto min = args.min ? std::make_optional<int64_t>(std::stol(*args.min))
```
This should be irrelevant following a change that takes all integral configs as a numeric type.

NyaliaLui commented on src/v/config/constraint_types.h:
```diff
@@ -0,0 +1,155 @@
+     * enabled but not both. Currently, the validator enforces this but that's a
+     * soft enforcement.
+     */
+    std::optional<ss::sstring> min;
+    std::optional<ss::sstring> max;
```
The reason for the string conversions was to address issues at the byte boundaries between integral data types. However, I may have come up with an alternative and will include in the next push.

NyaliaLui commented on src/v/config/configuration.cc:
```diff
@@ -708,7 +722,10 @@ configuration::configuration()
+      7 * 24h,
+      property<std::optional<std::chrono::milliseconds>>::noop_validator,
+      std::nullopt,
+      retention_ms_constraint_methods)
```
> Can we consider making all of the constructor take a plain struct so we can name all the fields and what they mean?

I took a stab at this in https://github.com/NyaliaLui/redpanda/commit/0e63c60df450171bb06525624406d862ddc9632f and made a ticket for this at https://github.com/redpanda-data/redpanda/issues/14621. This is a QoL improvment but it could be really helpful.

NyaliaLui replied:
Reordered constraint_methods that minimizes the blast radius but also retains argument context.

rockwotj replied:
Thanks, honestly this would have been good opportunity to add a new constructor that took an args struct so we could use named args for almost everything here.

NyaliaLui replied:
Wrapping config params in a struct is a great QoL improvement but it is out-of-scope of this project.

NyaliaLui commented on src/v/config/constraint_types.cc:
```diff
@@ -0,0 +1,176 @@
+  : _args{std::move(other._args)}
+  , _methods{std::move(other._methods)} {}
+
+constraint::constraint(struct constraint_args args, constraint_methods methods)
```
This is dead code from a previous iteration where the compiler complained about type hiding with structs. I will remove in the next push.

NyaliaLui commented on src/v/config/constraint_types.h:
```diff
@@ -0,0 +1,155 @@
+class constraint {
+public:
+    constraint() = default;
+    constraint(struct constraint_args, constraint_methods);
```
This is dead code from a previous iteration where the compiler complained about type hiding with structs. I will remove in the next push.

NyaliaLui commented on src/v/config/constraint_store.h:
```diff
@@ -0,0 +1,74 @@
+    }
+
+private:
+    std::unordered_map<std::string_view, constraint> _constraints;
```
> A btree

The idea was to leverage the search and insert runtime advantages that unordered_map has over map and btree impls. Also would not need to check for duplicates.

> A flat_hash_map

The benefits to flat_hash_map are memory efficiency and invalidation to iterators, references, and pointers such as upon rehash. This impl does not store the key or value as references or pointers so we can ignore those.

The motivation for this project is to assist users with ensuring data safety on Kafka topics so the expectation is that constraints are set infrequently: at startup from reading the YAML config file or with the Admin API. Unless a user is invoking the Admin API to dynamically change constraints in response to events, which is a strange use-case, the benefits for memory efficiency and iterator invalidation are negligible.

With that said, a possible corner case for frequent changes with the Admin API is at early feature adoption because users are figuring out optimal constraints. This also implies that, after some time, constraints are unchanged.

Happy to chat more offline if this requires further discussion.

NyaliaLui replied:
After giving this more thought, a flat_hash_map is better since the map is rehashed when loading constraints.

NyaliaLui commented:
... cherry picks didn't resolve conflicts ... will rebase dev another time.

Changes from [force-push](https://github.com/redpanda-data/redpanda/commit/40fe47622b242568794db297c06758e048724cc8):
- Improve comments
- Switched to absl::flat_hash_map
- Move code around to match commits and split up headers into impl files
- Reorded config arguments to retain argument context
- Using a variant<signed int, unsigned int, bool>
- Removed string conversions
- De-duplicated a lot of code
- Matched rjson_serialize to yaml decoding
- Removed none constraint type
- Switched to GTEST
- Switched to ssx::sformat
- Removed dead code
- Added noexcept to some methods
- Use insert_or_assign
- Using list initialization
- Prefer inline constexpr
- Removed make_range method
- Prefer get() to get0()
- Removed factory method on tristate

rockwotj reviewed (commented):
Can you rebase to drop Ben's OIDC commits before I review? Thanks!

NyaliaLui commented:
Changes from [force-push](https://github.com/redpanda-data/redpanda/commit/35a8723ef835e8e5bba78cc9628d64114b62f944):
- Rebase dev

NyaliaLui commented on src/v/config/constraint_store.h:
```diff
@@ -0,0 +1,46 @@
+#include <absl/container/flat_hash_map.h>
+
+namespace config {
+/**
+ * A store to track configuration constraints. This assists with de-coupling
+ * constraint arguments from their methods that are used to validate or clamp
+ * topic properties. Additionally, adding and searching for constraints is
+ * easier.
+ */
+class constraint_store {
```
This could be replaced with several helper methods for searching and inserting constraints. Created a follow-up ticket here https://github.com/redpanda-data/redpanda/issues/14774

NyaliaLui commented on tests/rptest/tests/cluster_config_test.py:
```diff
@@ -1862,3 +1863,94 @@ def test_legacy_default_explicit_after_upgrade(self, wipe_cache: bool):
+    RETENTION_MS_MAX = RETENTION_MS * 2
+    topics = [TopicSpec(retention_ms=RETENTION_MS)]
+
+    LOG_LIFETIME_CONSTRAINT = {
```
typo
```suggestion
LOG_RETENTION_CONSTRAINT = {
```

mattschumpert commented:
You can remove me from the code review list @NyaliaLui

michael-redpanda commented on src/v/config/constraint_method_generators.cc:
```diff
@@ -0,0 +1,561 @@
+           || greater_than_max<RangeT, T>(range.max, topic_val);
+}
+
+template<typename RangeT>
+range_values<RangeT>
+get_range(const constraint_args& args, const model::topic& topic) {
+    // NOTE to reviewers: Not using visit here or else need to check for
+    // signed/unsigned int comparison which is avoidable.
+    try {
+        return std::get<range_values<RangeT>>(args.flags);
+    } catch (const std::bad_variant_access& ex) {
+        // This should not happen since the variant is hardcoded in
+        // constraint_types.cc, but just-in-case, let's catch and log.
+        logger.error(
+          "Constraints failure[min/max not found]: topic property {}.{}",
+          topic(),
+          args.name);
+        return range_values<RangeT>{};
+    }
+}
```
nit/suggestion: Maybe return a `std::optional`?

NyaliaLui replied:
Removed in the recent push.

michael-redpanda commented on src/v/config/constraint_method_generators.cc:
```diff
@@ -0,0 +1,561 @@
+bool match_cluster_property(
+  const T& topic_val,
+  const T& cluster_val,
+  const std::variant<
+    range_values<int64_t>,
+    range_values<uint64_t>,
+    constraint_enabled_t>& flags) {
```
suggestion: it may be helpful to define this as a type in case the `variant` changes

michael-redpanda commented on src/v/config/constraint_method_generators.cc:
```diff
@@ -0,0 +1,561 @@
+    range_values<int64_t>,
+    range_values<uint64_t>,
+    constraint_enabled_t>& flags) {
+    return ss::visit(
+      flags,
+      [](const range_values<int64_t>) { return false; },
+      [](const range_values<uint64_t>) { return false; },
+      [&topic_val, &cluster_val](const constraint_enabled_t enabled) {
+          return enabled == constraint_enabled_t::yes
+                 && topic_val == cluster_val;
+      });
```
I think you may be able to do something like this:

```suggestion
    return ss::visit(
      flags,
      [&topic_val, &cluster_val](const constraint_enabled_t enabled) {
          return enabled == constraint_enabled_t::yes
                 && topic_val == cluster_val;
      },
     [](auto&&) { return false; });
```

Could be wrong

michael-redpanda commented on src/v/config/constraint_method_generators.cc:
```diff
@@ -0,0 +1,561 @@
+}
+
+template<typename T>
+bool match_cluster_property(
```
Ok I think I see what's going on with this function.  This validates if the value of the topic property matches the value of the cluster property.  nit: Maybe change name of function to like `match_cluster_property_value`?

michael-redpanda commented on src/v/config/constraint_method_generators.cc:
```diff
@@ -0,0 +1,561 @@
+  const T& cluster_val,
+  const model::topic& topic,
+  const constraint_args& args) {
+    if (!match_cluster_property(topic_val, cluster_val, args.flags)) {
+        logger.error(
+          "Constraints failure[does not match the cluster default {}]: "
+          "topic property {}.{}, value {}",
+          cluster_val,
+          topic(),
+          args.name,
+          topic_val);
+        return false;
+    }
```
I think this errors if `constraint_enabled_t` value is false (no), correct?  Is this the correct behavior?  If it isn't enabled wouldn't this succeed?

NyaliaLui replied:
Addressed the bug in the latest push.

michael-redpanda commented on src/v/config/constraint_method_generators.cc:
```diff
@@ -0,0 +1,561 @@
+    }
+}
+
+template<typename TopicFunc, typename ClusterFunc>
+constraint_validator_t
+make_validator(TopicFunc&& get_topic_val, ClusterFunc&& get_cluster_val) {
+    return [get_topic_val{std::forward<TopicFunc>(get_topic_val)},
+            get_cluster_val{std::forward<ClusterFunc>(get_cluster_val)}](
+             const cluster::topic_configuration& topic_cfg,
+             const constraint_args& args) {
+        auto topic_val = get_topic_val(topic_cfg);
+        return ternary_property_op(
+          config::shard_local_cfg().get(args.name),
+          [&topic_val, &topic_cfg, &args] {
+              return validate_range<int64_t>(
+                topic_val, topic_cfg.tp_ns.tp, args);
+          },
+          [&topic_val, &topic_cfg, &args] {
+              return validate_range<uint64_t>(
+                topic_val, topic_cfg.tp_ns.tp, args);
+          },
+          [&topic_val, &topic_cfg, &args, &get_cluster_val] {
+              // For property with enabled constraint
+              return validate_property(
+                topic_val, get_cluster_val(), topic_cfg.tp_ns.tp, args);
+          });
+    };
+}
+
+template<typename TopicFunc>
+constraint_validator_t make_validator(TopicFunc&& get_topic_val) {
+    return make_validator(
+      get_topic_val, [get_cluster_val{std::forward<TopicFunc>(get_topic_val)}] {
+          return get_cluster_val(cluster::topic_configuration{});
+      });
+}
```
I'm a little lost trying to parse this out.  In which situations would you use the `make_validator`? This is to lock a topic config to a cluster config, right?

michael-redpanda replied:
I think it's the parameter naming that's throwing me here

NyaliaLui replied:
Removed.

michael-redpanda commented on src/v/config/constraint_method_generators.cc:
```diff
@@ -0,0 +1,561 @@
+}
+
+namespace {
+// Temporary replacement for model::shadow_indexing_mode::drop_fetch
```
why temporary?

michael-redpanda reviewed (commented):
I know we talked about this, but I think it would be helpful to give an example of how these constraints are used.  It's unclear to me how things determined if they are clamped or restricted and how it all works.

Also I think we need either more comments or a more detailed cover letter going into what is happening in the PR.  There are a lot of new types introduced with very similar names.

NyaliaLui commented on src/v/config/constraint_types.h:
```diff
@@ -0,0 +1,175 @@
+} // namespace cluster
+
+namespace config {
+enum class constraint_type {
+    restrikt = 0,
+    clamp = 1,
+};
```
In our tree there are examples following the numbered pattern such as `config_resource_operation` and examples that do not follow the numbered pattern such as `base_property::visibility`. So the style is not consistent and I am a little confused where this comment is coming from, could you elaborate? This seems like a matter of preference.

michael-redpanda replied:
You rightly point out that the use of an enum base is inconsistent within the code base.  As for the numbers, they are redundant as those are the values guaranteed by the C++ spec.  It's a nit.

NyaliaLui commented on src/v/config/constraint_types.cc:
```diff
@@ -0,0 +1,209 @@
+      [&node, &rhs] { return maybe_set_range<int64_t>(node, rhs); },
+      [&node, &rhs] { return maybe_set_range<uint64_t>(node, rhs); },
+      [&node, &rhs] {
+          // For "enabled" contraints
+          if (node["enabled"] && !node["enabled"].IsNull()) {
+              rhs.flags = config::constraint_enabled_t(
+                node["enabled"].as<bool>());
+              return true;
+          }
+
+          return false;
```
In your scenario, enabled is read as Null by the YAML parser and parsing would fail and a message is automatically logged. In other comments I provided examples of what the error looks like. I also included sections for YAML parsing failures in the RFC.

NyaliaLui commented on src/v/config/constraint_method_generators.cc:
```diff
@@ -0,0 +1,561 @@
+template<typename RangeT, typename T>
+bool validate_range(
+  const T& topic_val, const model::topic& topic, const constraint_args& args) {
+    range_values<RangeT> range = get_range<RangeT>(args, topic);
```
Removed in the recent push.

NyaliaLui commented on src/v/config/constraint_types.cc:
```diff
@@ -0,0 +1,209 @@
+      [&node, &rhs] { return maybe_set_range<int64_t>(node, rhs); },
+      [&node, &rhs] { return maybe_set_range<uint64_t>(node, rhs); },
+      [&node, &rhs] {
+          // For "enabled" contraints
+          if (node["enabled"] && !node["enabled"].IsNull()) {
+              rhs.flags = config::constraint_enabled_t(
+                node["enabled"].as<bool>());
+              return true;
+          }
```
No, enabled constraints are constraints that check the topic-level property against the corresponding cluster-level one. This is illustrated in the RFC for non-integral properties.

NyaliaLui commented on src/v/config/constraint_method_generators.cc:
```diff
@@ -0,0 +1,561 @@
+    auto get_topic_val = [](const cluster::topic_configuration& topic_cfg) {
+        // TODO(@NyaliaLui): Make core thread proposing a default value for
+        // cluster-level retention local target bytes. Hardcoding 0 value is
+        // hazardous.
```
Removed

NyaliaLui commented:
Changes from [force-push](https://github.com/redpanda-data/redpanda/commit/2414f6af02184a6264ee594e55931c510e6271d4) and [force-push](https://github.com/redpanda-data/redpanda/commit/92b9e7e27bebe6d56fff13296048806d6a1df64c) and [force-push](https://github.com/redpanda-data/redpanda/commit/8d812bc8e78a91ebbdc0f0dc79884d3c6410dfbb):
- Removed confusing constraint args and methods
- Cleaned-up a lot of the duplicate code
- Decode YAML directly
- Catch bad optional access in YAML decoding
- Visit the variant with auto where possible

vbotbuildovich commented:
<!--NEW FAILURES-->
new failures in https://buildkite.com/redpanda/redpanda/builds/41029#018bcad8-fbaf-4fcd-87bd-26fe1b5bc70b:
```
"rptest.tests.cluster_config_test.ClusterConfigTest.test_valid_settings"
```
<!--8aa58a62-df7a-4346-90e6-56740be5205d-->

NyaliaLui commented:
Changes from [force-push](https://github.com/redpanda-data/redpanda/commit/83f31e4acf96863263afc0b9c9dec1c123f1b159):
- Add codegen'd kafka headers to the config CMakeLists

dotnwat commented on src/v/config/constraint_types.cc:
```diff
@@ -0,0 +1,703 @@
+          partition_count_opt,
+          constraint,
+          topic_cfg.tp_ns.tp,
+          kafka::topic_property_partition_count,
```
The `config` library is rather foundational in Redpanda--everything includes it. I think we should catch this reverse dependency (along with the need for the kafka codegen to bring in the request types) before it gets baked in and becomes very hard to untangle later.

michael-redpanda commented on src/v/config/constraints.cc:
```diff
@@ -45,6 +52,500 @@ std::ostream& operator<<(std::ostream& os, const constraint_t& constraint) {
+          constraint,
+          topic_cfg.tp_ns.tp,
+          kafka::topic_property_partition_count,
+          config::shard_local_cfg().default_topic_partitions.name())
```
question: sorry maybe I'm confused or I'm missing something obvious, but why isn't the value of `default_topic_partitions` passed in?  Can the constraint on the partition count only be a range?

NyaliaLui replied:
> Can the constraint on the partition count only be a range?

Correct. Integral properties get constraints with a range and non-integral properties get "enabled" constraints (perhaps a better name is match_cluster_val_constraint). This is why the cluster value for partition count is **not** passed in. Additionally, YAML parsing will fail if a user tries to use "enabled" for an integral property or "min/max" for a non-integral property.

michael-redpanda commented on src/v/config/constraints.cc:
```diff
@@ -45,6 +52,500 @@ std::ostream& operator<<(std::ostream& os, const constraint_t& constraint) {
+    auto replication_factor_opt = std::make_optional(
+      topic_cfg.replication_factor);
+    bool ret
+      = validate_value(
+          partition_count_opt,
+          constraint,
+          topic_cfg.tp_ns.tp,
+          kafka::topic_property_partition_count,
+          config::shard_local_cfg().default_topic_partitions.name())
```
Since you know the name of the constraint, can you do a lookup to get the related cluster config rather than having to check every one?  Because only one cluster config is going to match the provided constraint, correct?

michael-redpanda replied:
Or is this large conditional predicated on trying to extract the correct topic property?

michael-redpanda replied:
Could you write out a switch/if statement that checks the name in `constraint`?

NyaliaLui replied:
So, earlier iterations had the switch/if statement that you are referring to but there were a couple "clean-code" problems with it, such as: (1) It's a giant switch/if. One for each property name (2) `topic_config_satisfies_constraint` and `constraint_clamp_topic_config` would both need the giant switch/if. De-duplicating the code led to some cascading changes for calling validate_value vs clamp_value and optional vs tristate which also looked un-clean. So I opted for the current approach. I am happy to re-visit this approach if it is a larger concern.

NyaliaLui replied:
A shorter answer: the constraint name is checked within `validate_value`, so the correct configs are checked against their corresponding constraint and it looks nicer than a big if statement.

michael-redpanda commented on src/v/config/constraints.cc:
```diff
@@ -45,6 +52,500 @@ std::ostream& operator<<(std::ostream& os, const constraint_t& constraint) {
+          topic_cfg.tp_ns.tp,
+          kafka::topic_property_segment_size,
+          config::shard_local_cfg().log_segment_size.name())
+        || validate_value(
+          topic_cfg.properties.retention_bytes,
+          constraint,
+          topic_cfg.tp_ns.tp,
+          kafka::topic_property_retention_bytes,
+          config::shard_local_cfg().retention_bytes.name())
+        || validate_value(
+          topic_cfg.properties.retention_duration,
+          constraint,
+          topic_cfg.tp_ns.tp,
+          kafka::topic_property_retention_duration,
+          config::shard_local_cfg().log_retention_ms.name())
```
question: how do you handle a situation for these properties where `-1` is a valid value but represents a positive infinite number?

Scenario: Customer sets a minimum log retention time to be 7 days, but no maximum.  User creates topic with log retention ms to be '-1'.  That would fail validation, correct?

NyaliaLui replied:
The problem was with when to apply -1 for clamp constraints since a clamp is applied when the current value is less than the minimum. But Redpanda assigns -1 (disabled state) at topic creation and -1 is never less than -1. So I went with a design that fails validation here because -1 is not within any range [0, inf+). Of course this will need to be documented.

NyaliaLui replied:
Asked about this in [slack thread](https://redpandadata.slack.com/archives/C04AVDG7N58/p1700167194539089) because depending on semantics, validation could have different results.

NyaliaLui commented on src/v/config/constraints.cc:
```diff
@@ -45,6 +52,500 @@ std::ostream& operator<<(std::ostream& os, const constraint_t& constraint) {
+        return true;
+    }
+
+    // A constraint with "enabled" flag means that the topic property must match
+    // the cluster one. An undefined topic value or cluster value could not
+    // match the other, this implies that the constraint is not satisfied.
+    if (!topic_val || !cluster_val) {
+        return false;
+    }
```
Cluster value is defined for non-integral constraints and control does **not** get here for integral constraint. This is hard coded in. The check here is in-case someone in the future passes an undefined cluster value.

NyaliaLui commented on src/v/config/constraints.cc:
```diff
@@ -45,6 +52,500 @@ std::ostream& operator<<(std::ostream& os, const constraint_t& constraint) {
+  const model::topic& topic,
+  const std::string_view& topic_property) {
+    if (!(valid_min(topic_val, range.min) && valid_max(topic_val, range.max))) {
+        vlog(
+          constraints_log.error,
+          "Constraints failure[value out-of-range]: topic property {}.{}, "
+          "value {}",
+          topic(),
+          topic_property,
+          topic_val);
```
IIUC, systems that use constraints such as Kafka lack sufficient context to determine if the log message should indicate the range or enabled flags. There is, however, enough context within the constraints system.

NyaliaLui commented:
Changes from [force-push(https://github.com/redpanda-data/redpanda/commit/43411f04f9335f1aa7d73954a1324c1eaddf22d4) and [force-push](https://github.com/redpanda-data/redpanda/commit/5c3769dbbb4a06d538fccf00f688e8f42b79cd05):
- Check for disabled state and, depending on the property in question, this state represents infinity or off

NyaliaLui commented on src/v/config/constraints.cc:
```diff
@@ -45,6 +54,531 @@ std::ostream& operator<<(std::ostream& os, const constraint_t& constraint) {
+        // minimum or it could mean turned-off which is never within a range.
+        if (is_infinite(topic_val, topic_property)) {
+            return true;
+        } else if (is_turned_off(topic_val, topic_property)) {
+            return false;
+        }
```
Thanks for the catch.

michael-redpanda commented on src/v/config/property.h:
```diff
@@ -702,6 +710,31 @@ bool property<T>::is_array() const {
     return detail::is_array<T>();
 }
 
+template<typename T>
+bool property<T>::is_signed() const {
+    return std::is_signed_v<T>;
+}
+
+template<typename T>
+bool property<T>::is_unsigned() const {
+    return std::is_unsigned_v<T>;
+}
+
+template<typename T>
+bool property<T>::is_bool() const {
+    return std::is_same_v<T, bool>;
+}
```
question: do these properly handle configurations that are, for example `std::optional<size_t>`?  Will `is_unsigned()` return true for `retention_bytes` which is such a type?

michael-redpanda commented on src/v/config/constraints.cc:
```diff
@@ -0,0 +1,727 @@
+}
+} // namespace
+
+bool topic_config_satisfies_constraint(
+  const cluster::topic_configuration& topic_cfg,
+  const constraint_t& constraint) {
```
suggestion: Why not have the caller supply the actual topic config value and cluster config?  Something like:

```
template<typename T>
bool topic_config_satisifes_constraint(
  T & topic_config_val
  const property<T> &cluster_config)
```

This may need a few specializers maybe for `std::optional` topic config values

You can obtain the name of the cluster property via `cluster_config.name()` and with that you can look up the corresponding constraint, if it exists.  You may still need a map of sorts to get the corresponding topic property name

It looks like the caller already needs to know the name of the constraint in order to access it from the constraint map, right?

NyaliaLui replied:
> It looks like the caller already needs to know the name of the constraint in order to access it from the constraint map, right?

We do not need to know the name of the constraint in the general case. [Here](https://github.com/NyaliaLui/redpanda/commit/3065a563641b06ee3a91c2fa2b5b06dd466103b1#diff-2526100bdb158053e8146a22623b4fa084c33b8730e47715c43028636902235aR303-R317) is an example of what using this API looks like for create_topics. Note that `apply_constraint` was renamed to `valid_topic_config` (I'm still working on the naming but that is a non-blocker for this PR). Notice that the broker iterates over every constraint and then checks the topic configuration against the constraint.

For those APIs that need a specific constraint, a search by name is simple. [Here](https://github.com/NyaliaLui/redpanda/commit/3065a563641b06ee3a91c2fa2b5b06dd466103b1#diff-215eea521e18fe56d728c5bf64ada99aa4790c0d7f0816b660a2b6f156b99fd3R286-R301) is an example with CreatePartitions.

> Why not have the caller supply the actual topic config value and cluster config?

Having the caller supply the actual topic config value would re-create a similar structure to what is already included. [here](https://github.com/NyaliaLui/redpanda/commit/19039107db4b229a1135049c327bd77f6d15566d#diff-c618e718c29ebdabd2aadde150ff01dc2b6750369c9f798dec0c1de182791459R364-R378) is an example at the callsite.

> You can obtain the name of the cluster property via cluster_config.name() and with that you can look up the corresponding constraint, if it exists.

See the above comment about constraint search.

BenPope commented on src/v/config/constraints.cc:
```diff
@@ -45,6 +54,534 @@ std::ostream& operator<<(std::ostream& os, const constraint_t& constraint) {
+
+void constraint_clamp_topic_config(
+  cluster::topic_configuration& topic_cfg, const constraint_t& constraint) {
+    auto partition_count_tri = tristate(
+      std::make_optional(topic_cfg.partition_count));
+    auto replication_factor_tri = tristate(
+      std::make_optional(topic_cfg.replication_factor));
```
It's not obvious to me why these ones are separate. I see a few cases:
* The topic property is a tristate, it's taken by reference and modified
* The topic property is an optional, an intermediate tristate is constructed, and the optional is written back.
* The topic property is is remote_delete - it's ignored
* Oh, I see, these are topic configs!
  * Is it worth hiding these write-backs inside a clamp_value that takes a scaler and wraps it in an optional, ensuring it's written back?

BenPope commented on src/v/config/constraints.cc:
```diff
@@ -45,6 +54,534 @@ std::ostream& operator<<(std::ostream& os, const constraint_t& constraint) {
+          kafka::topic_property_remote_write,
+          config::shard_local_cfg().cloud_storage_enable_remote_write.name(),
+          config::shard_local_cfg().cloud_storage_enable_remote_write());
+        topic_cfg.properties.shadow_indexing
+          = model::get_shadow_indexing_mode_impl(
+            *archival_enabled_opt, *fetch_enabled_opt);
```
I guess we can't easily ensure this one is written back.

BenPope commented on src/v/config/constraints.cc:
```diff
@@ -45,6 +54,534 @@ std::ostream& operator<<(std::ostream& os, const constraint_t& constraint) {
+      kafka::topic_property_timestamp_type,
+      config::shard_local_cfg().log_message_timestamp_type.name(),
+      config::shard_local_cfg().log_message_timestamp_type());
+    clamp_value(
+      topic_cfg.properties.segment_size,
+      constraint,
+      topic_cfg.tp_ns.tp,
+      kafka::topic_property_segment_size,
+      config::shard_local_cfg().log_segment_size.name());
```
It took me a little while, but I checked that these are all essentially the same for `constraint_clamp_topic_config` and `topic_config_satisfies_constraint`.

I wonder if it would be worth extracting these properties into traits classes or something similar, so that they can be reused.

BenPope commented on src/v/config/constraints.h:
```diff
@@ -0,0 +1,116 @@
+
+using constraint_enabled_t = ss::bool_class<struct constraint_enabled_tag>;
+
+namespace {
```
Do not use unnamed namespaces in header files _clang-tidy_ [cert-dcl59-cpp](https://clang.llvm.org/extra/clang-tidy/checks/cert/dcl59-cpp.html)

NyaliaLui commented on src/v/config/validators.cc:
```diff
@@ -183,4 +183,34 @@ validate_audit_event_types(const std::vector<ss::sstring>& vs) {
+  const std::unordered_map<ss::sstring, config::constraint_t>& constraint_t) {
+    std::optional<ss::sstring> msg{std::nullopt};
+    for (const auto& [name, args] : constraint_t) {
+        msg = ss::visit(
```
This does not accumulate, instead the last violation is reported. I can switch to the first if that is preferred.
